File: server/internal/handler/dto/sdo/billing.go
package sdo

// PlanFeatureDTO is one row of plan_features.
type PlanFeatureDTO struct {
	FeatureKey string `json:"feature_key" example:"members.max"`
	Enabled    bool   `json:"enabled" example:"true"`
	QuotaLimit *int64 `json:"quota_limit" description:"null = không giới hạn; 0 = tắt" example:"50"`
}

type PlanDTO struct {
	ID            string           `json:"id" example:"01K4F02PLAN0STARTER000000A"`
	Code          string           `json:"code" example:"starter"`
	Name          string           `json:"name" example:"Starter"`
	Description   string           `json:"description" example:"Gói mặc định cho mọi tổ chức"`
	BillingPeriod string           `json:"billing_period" description:"month, year hoặc none" example:"none"`
	PriceAmount   *int64           `json:"price_amount" description:"Đơn vị nhỏ nhất (VND: đồng); null = liên hệ" example:"0"`
	PriceCurrency string           `json:"price_currency" example:"VND"`
	IsDefault     bool             `json:"is_default" example:"true"`
	Features      []PlanFeatureDTO `json:"features"`
}

// PlanListSDO is GET /api/v1/plans.
type PlanListSDO struct {
	Plans []PlanDTO `json:"plans"`
}

type SubscriptionDTO struct {
	ID                 string `json:"id" example:"01K4F02SUB0N1P2Q3R4S5T6U7V"`
	PlanCode           string `json:"plan_code" example:"starter"`
	PlanName           string `json:"plan_name" example:"Starter"`
	Status             string `json:"status" description:"trialing, active, past_due, suspended hoặc canceled" example:"active"`
	Provider           string `json:"provider" description:"manual, stripe hoặc payos" example:"manual"`
	CurrentPeriodStart string `json:"current_period_start" example:"2026-09-06T00:00:00Z"`
	CurrentPeriodEnd   string `json:"current_period_end,omitempty" example:"2026-10-06T00:00:00Z"`
	CancelAt           string `json:"cancel_at,omitempty" example:"2026-10-06T00:00:00Z"`
	TrialEndsAt        string `json:"trial_ends_at,omitempty" example:"2026-09-20T00:00:00Z"`
	RowVersion         int32  `json:"row_version" example:"1"`
}

// EntitlementDTO is one feature as it applies to the organization now.
type EntitlementDTO struct {
	FeatureKey   string `json:"feature_key" example:"members.max"`
	Name         string `json:"name" example:"Thành viên tổ chức"`
	Kind         string `json:"kind" description:"flag hoặc quota" example:"quota"`
	Unit         string `json:"unit,omitempty" example:"members"`
	Category     string `json:"category" example:"organization"`
	Enabled      bool   `json:"enabled" example:"true"`
	QuotaLimit   *int64 `json:"quota_limit" description:"null = không giới hạn" example:"50"`
	CurrentUsage int64  `json:"current_usage" example:"12"`
}

// SubscriptionSDO is GET /api/v1/orgs/{orgID}/billing and the plan mutations.
type SubscriptionSDO struct {
	Subscription SubscriptionDTO  `json:"subscription"`
	Entitlements []EntitlementDTO `json:"entitlements"`
}

// CheckoutSDO is POST /api/v1/orgs/{orgID}/billing/checkout.
type CheckoutSDO struct {
	URL string `json:"url" example:"https://pay.example.com/session/abc"`
}

File: packages/core/types/billing.ts
import { z } from "zod";

export const SUBSCRIPTION_STATUSES = ["trialing", "active", "past_due", "suspended", "canceled"] as const;
export type SubscriptionStatus = (typeof SUBSCRIPTION_STATUSES)[number];

/** Lenient: server enums stay strings; the exported types narrow them. */
export const PlanFeatureSchema = z.object({
  feature_key: z.string(),
  enabled: z.boolean(),
  quota_limit: z.number().nullable().optional().default(null),
});
export type PlanFeature = z.infer<typeof PlanFeatureSchema>;

export const PlanSchema = z.object({
  id: z.string(),
  code: z.string(),
  name: z.string(),
  description: z.string().optional().default(""),
  billing_period: z.string().optional().default("none"),
  price_amount: z.number().nullable().optional().default(null),
  price_currency: z.string().optional().default("VND"),
  is_default: z.boolean().optional().default(false),
  features: z.array(PlanFeatureSchema).optional().default([]),
});
export type Plan = z.infer<typeof PlanSchema>;

export const SubscriptionSchema = z.object({
  id: z.string(),
  plan_code: z.string(),
  plan_name: z.string(),
  status: z.string(),
  provider: z.string(),
  current_period_start: z.string(),
  current_period_end: z.string().optional(),
  cancel_at: z.string().optional(),
  trial_ends_at: z.string().optional(),
  row_version: z.number(),
});
export type Subscription = Omit<z.infer<typeof SubscriptionSchema>, "status"> & {
  status: SubscriptionStatus;
};

/** One feature as it applies to the organization now, usage included. */
export const EntitlementSchema = z.object({
  feature_key: z.string(),
  name: z.string(),
  kind: z.string(),
  unit: z.string().optional(),
  category: z.string().optional().default("general"),
  enabled: z.boolean(),
  quota_limit: z.number().nullable().optional().default(null),
  current_usage: z.number().optional().default(0),
  /** false = nothing reads or writes this feature yet; the UI hides it. */
  metered: z.boolean().optional().default(true),
});
export type Entitlement = z.infer<typeof EntitlementSchema>;

export interface SubscriptionView {
  subscription: Subscription;
  entitlements: Entitlement[];
}
